core/io/multiplayer_api.cpp

```diff
@@ -1196,19 +1230,122 @@ Error MultiplayerAPI::_send_spawn_despawn(int p_peer_id, const ResourceUID::ID &
 
 	const CharString cname = String(names[names.size() - 1]).utf8();
 	int nlen = encode_cstring(cname.get_data(), nullptr);
-	MAKE_ROOM(1 + 8 + 4 + 4 + nlen + p_data_len);
+	MAKE_ROOM(1 + 8 + 4 + 4 + nlen + state_len);
 	uint8_t *ptr = packet_cache.ptrw();
-	ptr[0] = p_spawn ? NETWORK_COMMAND_SPAWN : NETWORK_COMMAND_DESPAWN;
+	ptr[0] = (p_spawn ? NETWORK_COMMAND_SPAWN : NETWORK_COMMAND_DESPAWN) + ((is_raw ? 1 : 0) << BYTE_ONLY_OR_NO_ARGS_SHIFT);
 	int ofs = 1;
 	ofs += encode_uint64(p_scene_id, &ptr[ofs]);
 	ofs += encode_uint32(psc->id, &ptr[ofs]);
 	ofs += encode_uint32(nlen, &ptr[ofs]);
 	ofs += encode_cstring(cname.get_data(), &ptr[ofs]);
-	memcpy(&ptr[ofs], p_data, p_data_len);
+	if (state_len) {
+		switch (state_type) {
+			case Variant::OBJECT:
+				_encode_spawn_state(cfg, p_state.get_validated_object(), &ptr[ofs], state_len, p_optimize_objects ? &is_raw : nullptr);
+				break;
+			case Variant::PACKED_BYTE_ARRAY: {
+				const PackedByteArray pba = p_state.operator PackedByteArray();
+				memcpy(&ptr[ofs], pba.ptr(), pba.size());
+			} break;
+			default:
+				ERR_FAIL_V(ERR_BUG);
+		}
+	}
 	network_peer->set_target_peer(p_peer_id);
 	network_peer->set_transfer_channel(0);
 	network_peer->set_transfer_mode(MultiplayerPeer::TRANSFER_MODE_RELIABLE);
-	return network_peer->put_packet(ptr, ofs + p_data_len);
+	return network_peer->put_packet(ptr, ofs + state_len);
+}
+
+PackedByteArray MultiplayerAPI::spawnable_encode_state(const ResourceUID::ID &p_scene_id, const Object *p_obj) {
+	PackedByteArray state;
+	ERR_FAIL_COND_V_MSG(!spawnables.has(p_scene_id), state, vformat("Spawnable not found: %d", p_scene_id));
+	const SpawnableConfig &cfg = spawnables[p_scene_id];
+	int len = 0;
+	Error err = _encode_spawn_state(cfg, p_obj, nullptr, len);
+	ERR_FAIL_COND_V_MSG(err != OK, state, "Unable to encode spawnable state.");
+	state.resize(len);
+	_encode_spawn_state(cfg, p_obj, state.ptrw(), len);
+	return state;
+}
+
+Error MultiplayerAPI::spawnable_decode_state(const ResourceUID::ID &p_scene_id, Object *p_obj, const PackedByteArray p_data) {
+	ERR_FAIL_COND_V_MSG(!spawnables.has(p_scene_id), ERR_INVALID_PARAMETER, vformat("Spawnable not found: %d", p_scene_id));
+	const SpawnableConfig &cfg = spawnables[p_scene_id];
+	int size;
+	return _decode_spawn_state(cfg, p_obj, p_data.ptr(), p_data.size(), size);
+}
+
+Error MultiplayerAPI::_encode_spawn_state(const SpawnableConfig &p_cfg, const Object *p_obj, uint8_t *p_buffer, int &r_len, bool *r_raw) {
+	r_len = 0;
+	ERR_FAIL_COND_V_MSG(!p_obj, ERR_INVALID_PARAMETER, "Cannot encode null object");
+
+	int size = 0;
+	bool valid = false;
+
+	// Try raw encoding optimization.
+	if (r_raw && p_cfg.properties.size() == 1) {
+		*r_raw = false;
+		const StringName &prop = p_cfg.properties[0];
+		const Variant v = p_obj->get(prop, &valid);
```

Comment:
This is, in fact, awful. I'm reading the properties 2 times for no real reason, will need to fix this